### PR TITLE
Add support for stable ids

### LINKED ISSUE #1371: Support Stable IDs
Support stable IDs as provided by upstream. Unfortunately, we cannot wait for a proper release of upstream packages for work on this to begin.

There should be no change in user-facing APIs or behavior.

Corresponding issue in realm-cocoa: https://github.com/realm/realm-cocoa/issues/4909
Corresponding issue in realm-java: https://github.com/realm/realm-java/issues/4542
Corresponding issue in realm-js: https://github.com/realm/realm-js/issues/985

Upstream PR: https://github.com/realm/realm-sync/pull/1170

Object Store PR: https://github.com/realm/realm-object-store/pull/434

### PR DESCRIPTION
**Pull request by nirinchev**
## Description

Fixes #1371 

##  TODO

* [ ] Changelog entry
* [x] Decide what to do with CreateObject
* [ ] Migration?
* [x] Revert `'windows && !nodejs'` to `'windows'` once https://github.com/realm/ci/issues/193 has been resolved.

### CODE REVIEW

File: wrappers/src/shared_realm_cs.cpp [hunk 1]
```diff
@@ -46,6 +46,39 @@ namespace binding {
     }
 }
     
+    template<typename KeyType>
+    Object* create_object_unique(SharedRealm* realm, Table* table, KeyType key, bool try_update, std::function<size_t (size_t)> find_first, std::function<void (size_t, size_t)> set_unique, std::string key_to_string, bool& is_new)
+    {
+        (*realm)->verify_in_write();
+        const std::string object_name(ObjectStore::object_type_for_table_name(table->get_name()));
```
Comment (simonask) [inline wrappers/src/shared_realm_cs.cpp]: Might as well use Core's `StringData` instead of `std::string` to avoid a heap allocation.

File: wrappers/src/shared_realm_cs.cpp [hunk 2]
```diff
@@ -46,6 +46,39 @@ namespace binding {
...
+        auto& object_schema = *realm->get()->schema().find(object_name);
```
Comment (simonask) [inline wrappers/src/shared_realm_cs.cpp]: Is this guaranteed, or can it throw if the schema does not contain the objet name?

File: wrappers/src/shared_realm_cs.cpp [hunk 3]
```diff
@@ -46,6 +46,39 @@ namespace binding {
...
+        
+        size_t column_index = object_schema.primary_key_property()->table_column;
+        size_t row_index = find_first(column_index);
+        
+        if (row_index == realm::not_found) {
+            is_new = true;
+#if REALM_ENABLE_SYNC && false
```
Comment (simonask) [inline wrappers/src/shared_realm_cs.cpp]: Remove `&& false`

File: wrappers/src/shared_realm_cs.cpp [hunk 4]
```diff
@@ -233,4 +266,59 @@ REALM_EXPORT void thread_safe_reference_destroy(ThreadSafeReferenceBase* referen
     delete reference;
 }
     
+REALM_EXPORT Object* shared_realm_create_object(SharedRealm* realm, Table* table, NativeException::Marshallable& ex)
+{
+    return handle_errors(ex, [&]() {
+        (*realm)->verify_in_write();
+        
+#if REALM_ENABLE_SYNC && false
```
Comment (simonask) [inline wrappers/src/shared_realm_cs.cpp]: Remove `&& false`

File: wrappers/src/shared_realm_cs.cpp [hunk 5]
```diff
@@ -46,6 +46,39 @@ namespace binding {
     }
 }
     
+    template<typename KeyType>
+    Object* create_object_unique(SharedRealm* realm, Table* table, KeyType key, bool try_update, std::function<size_t (size_t)> find_first, std::function<void (size_t, size_t)> set_unique, std::string key_to_string, bool& is_new)
```
Comment (simonask) [inline wrappers/src/shared_realm_cs.cpp]: Small detail, but it would probably be more efficient to use an opaque template parameter instead of `std::function<...>` here.

**Comment by nirinchev** [comment]
Regarding `sync::create_table_with_primary_key`, my understanding is that it's implemented in https://github.com/realm/realm-object-store/pull/434 and we'll get it by virtue of calling `RealmCoordinator::get_realm`. I hadn't updated the submodule reference because I expected your PR to get merged before a sync release with stable ids is cut.

File: wrappers/src/shared_realm_cs.cpp (hunk 2 above)
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp]: I'm not sure how we could get in a situation where the schema doesn't contain the object name.

File: wrappers/src/shared_realm_cs.cpp [hunk 6]
```diff
@@ -46,6 +46,39 @@ namespace binding {
     }
 }
     
+    template<typename KeyType, typename FindFirst>
+    Object* create_object_unique(SharedRealm* realm, Table* table, KeyType key, bool try_update, FindFirst find_first, std::function<void (size_t, size_t)> set_unique, std::string key_to_string, bool& is_new)
```
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp]: We can wait for sync release that includes the changes in Core 2.7 to use the generic `set_unique` and `find_first`: https://github.com/realm/realm-core/blob/master/CHANGELOG.md

**Comment by nirinchev** [comment]
This depends on https://github.com/realm/realm-sync/pull/1353 being merged

File: wrappers/src/shared_realm_cs.cpp [hunk 7]
```diff
@@ -259,3 +259,120 @@ REALM_EXPORT void shared_realm_write_copy(SharedRealm* realm, uint16_t* path, si
...
+{
+    const StringData object_name(ObjectStore::object_type_for_table_name(table.get_name()));
+    return *realm->schema().find(object_name);
+}
+
+namespace realm
+{
+// specialize overload
+template <>
+size_t Table::find_first(size_t column_ndx, null) const
```
Comment (bdash) [inline wrappers/src/shared_realm_cs.cpp]: Why is this necessary? Did https://github.com/realm/realm-core/pull/2669 not make it into a core release yet?

Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp]: As far as I know - not yet.

Comment (bdash) [inline wrappers/src/shared_realm_cs.cpp]: It looks like it's in core 3.0.0-rc2 at least, but sync hasn't switched over to that so I guess bindings can't take advantage of it.

File: wrappers/Makefile [hunk 8]
```diff
@@ -17,9 +17,9 @@ ifeq ($(REALM_ENABLE_SYNC),1)
 else
 	APPLE_PREPROCESSOR_DEFINITIONS=""
 	WRAPPERS_XCODE_TARGET=wrappers
-	APPLE_BUNDLE_NAME=realm-core-$(CORE_VER).tar.xz
+	APPLE_BUNDLE_NAME= realm-core-cocoa-v$(CORE_VER).tar.xz
 	APPLE_BUNDLE_URL="http://static.realm.io/downloads/core/$(APPLE_BUNDLE_NAME)"
```
Comment (fealebenpae) [inline wrappers/Makefile]: `http://static.realm.io/downloads/core/v$(CORE_VER)/cocoa/$(APPLE_BUNDLE_NAME)`

File: wrappers/Makefile [hunk 9]
```diff
@@ -17,9 +17,9 @@ ifeq ($(REALM_ENABLE_SYNC),1)
...
-	ANDROID_BUNDLE_NAME=realm-core-android-$(CORE_VER).tar.gz
+	ANDROID_BUNDLE_NAME=realm-core-android-v$(CORE_VER).tar.gz
 	ANDROID_BUNDLE_URL="http://static.realm.io/downloads/core/$(ANDROID_BUNDLE_NAME)"
```
Comment (fealebenpae) [inline wrappers/Makefile]: `http://static.realm.io/downloads/core/v$(CORE_VER)/android/$(ANDROID_BUNDLE_NAME)`

File: wrappers/Makefile (hunk 8 above)
Comment (nirinchev) [inline wrappers/Makefile]: In OS it appears there's a different calculation logic: https://github.com/realm/realm-object-store/blob/master/CMake/RealmCore.cmake#L196, so not sure which one I should use. Can you link me to script that does the uploading - I wasn't able to find it.

File: Realm/Realm/Handles/SharedRealmHandle.cs [hunk 10]
```diff
@@ -225,6 +245,72 @@ public void WriteCopy(string path, byte[] encryptionKey)
...
+        }
+
+        public IntPtr CreateObjectWithPrimaryKey(Property pkProperty, object primaryKey, TableHandle table, string parentType, bool update, out bool isNew)
+        {
+            switch (pkProperty.Type)
+            {
+                case PropertyType.String:
+                    if (primaryKey == null)
+                    {
+                        throw new ArgumentNullException(nameof(primaryKey), "Object identifiers cannot be null");
```
Comment (bdash) [inline Realm/Realm/Handles/SharedRealmHandle.cs]: Perhaps I'm misunderstanding the logic of this function, but it looks like it's rejecting null strings as primary keys, but is completely happy with null integers for primary keys? Is that intentional?

Comment (nirinchev) [inline Realm/Realm/Handles/SharedRealmHandle.cs]: Yes, we have never supported null strings as primary keys.

Comment (bdash) [inline Realm/Realm/Handles/SharedRealmHandle.cs]: It seems like a strange inconsistency to support null integers but not strings as primary keys. AFAIK other bindings support null strings as primary keys, so .NET likely needs to also for sake of synced Realms.

File: CHANGELOG.md [hunk 11]
```diff
@@ -13,6 +13,17 @@ per-application keychain service. ([#1522](https://github.com/realm/realm-dotnet
 ### Breaking Changes
 
 1.6.0 (2017-08-14)
+2.0.0 (TBD)
```
Comment (nirinchev) [inline CHANGELOG.md]: Looks like the changelog has been incorrectly merged - we should fix that.

File: Jenkinsfile [hunk 12]
```diff
@@ -421,10 +395,38 @@ stage('Test with sync') {
   )
 }
 
+def buildAndroidWrappers(abis, String stashName, Map extraCMakeArguments = [:]) {
```
Comment (nirinchev) [inline Jenkinsfile]: The `abis` we pass seem to be the same everywhere we use this, should we hardcode them instead?

File: Realm/Realm/Handles/SharedRealmHandle.cs [hunk 13]
```diff
@@ -225,6 +245,72 @@ public void WriteCopy(string path, byte[] encryptionKey)
...
+                    }
+
+                    if (primaryKey is string stringKey)
+                    {
+						return CreateObjectWithPrimaryKey(table, stringKey, update, out isNew);
```
Comment (nirinchev) [inline Realm/Realm/Handles/SharedRealmHandle.cs]: Tabs must be spaces

File: Realm/Realm/Handles/SharedRealmHandle.cs [hunk 14]
```diff
@@ -225,6 +245,72 @@ public void WriteCopy(string path, byte[] encryptionKey)
...
+                    return CreateObjectWithPrimaryKey(table, longKey, pkProperty.Type.IsNullable(), update, out isNew);
+                default:
+                    throw new NotSupportedException($"Unexpected primary key of type: {pkProperty.Type}");
+            }
+        }
+
+        private IntPtr CreateObjectWithPrimaryKey(TableHandle table, long key, bool isNullable, bool update, out bool isNew)
+        {
+            NativeException ex;
+            var result = NativeMethods.create_object_unique(this, table, key, isNullable, update, out isNew, out ex);
```
Comment (nirinchev) [inline Realm/Realm/Handles/SharedRealmHandle.cs]: Should be able to use `out var ex`

File: Tests/Tests.Shared/PrimaryKeyTests.cs [hunk 15]
```diff
@@ -103,22 +103,18 @@ public void FailToFindByPrimaryKeyDynamicTests(Type type, object primaryKeyValue
         [TestCase(typeof(PrimaryKeyStringObject), "key")]
         public void CreateObject_WhenPKExists_ShouldFail(Type type, object primaryKeyValue)
         {
-            var pkProperty = type.GetProperties().Single(p => p.GetCustomAttribute<PrimaryKeyAttribute>() != null);
-
             _realm.Write(() =>
             {
-                var first = _realm.CreateObject(type.Name);
-                pkProperty.SetValue(first, primaryKeyValue);
+                var first = _realm.CreateObject(type.Name, primaryKeyValue);
```
Comment (nirinchev) [inline Tests/Tests.Shared/PrimaryKeyTests.cs]: No need to assign anymore

File: Tests/Tests.Shared/PrimaryKeyTests.cs [hunk 16]
```diff
@@ -103,22 +103,18 @@ public void FailToFindByPrimaryKeyDynamicTests(Type type, object primaryKeyValue
...
             });
 
             Assert.That(() =>
             {
                 _realm.Write(() =>
                 {
-                    var second = _realm.CreateObject(type.Name);
-                    pkProperty.SetValue(second, primaryKeyValue);
+                    var second = _realm.CreateObject(type.Name, primaryKeyValue);
```
Comment (nirinchev) [inline Tests/Tests.Shared/PrimaryKeyTests.cs]: No need for the `var second`

File: wrappers/src/shared_realm_cs.cpp [hunk 17]
```diff
@@ -263,3 +263,104 @@ REALM_EXPORT void shared_realm_write_copy(SharedRealm* realm, uint16_t* path, si
...
+        std::ostringstream string_builder;
+        string_builder << key;
+        throw SetDuplicatePrimaryKeyValueException(object_schema.name,
+                                                   primary_key_property.name,
+                                                   string_builder.str());
+    } else {
+        is_new = false;
+    }
+
+
```
Comment (nirinchev) [inline wrappers/src/shared_realm_cs.cpp]: Extra empty line

File: wrappers/CMakeLists.txt [hunk 18]
```diff
@@ -111,8 +112,8 @@ set(PEGTL_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/src/object-store/external/pegt
 add_subdirectory(src/object-store/src)
 add_subdirectory(src)
 
-if(APPLE)
-  # force Object Store to use the generic EventLoopSignal implementation instead of the NSRunLoop-one
+if(APPLE OR ANDROID)
```
Comment (fealebenpae) [inline wrappers/CMakeLists.txt]: On Windows and Linux it already picks up the generic EventLoopSignal. We’d get linker errors if we compiled the file twice.

**Comment by fealebenpae** [comment]
@nirinchev we need a new sync 2.x release because 2.0-rc15 doesn’t include feature tokens.

**Comment by nirinchev** [comment]
This passes CI now.
